### ISSUE TITLE
Proposal: require `notifications.show_warning` instead of `warnings.warn` for GUI warnings (vote please)

### ISSUE THREAD

**Issue by tlambert03** [napari/napari#3787]
## 🧰 Task
We've discussed in multiple places now the possibility of no longer hooking into `warnings.warn` for napari gui notifications, but rather requiring the explicit napari API `notifications.show_warning`.  It was a nice "pythonic" idea, but it can be too noisy, and the fraction of warnings that actually need to be shown to the GUI user are probably low.

vote please:
👍 = require `show_warning` to surface a warning in the gui
👎 = keep using `warnings.warn` as the API to surface warnings in the GUI

**Comment by jni** [comment]
@tlambert03 to clarify: do you mean that e.g. the private attribute access would *not* be surfaced to the UI, only specific notifications from us/plugin devs would be surfaced, and everything else would go to the console?

**Comment by tlambert03** [comment]
> do you mean that e.g. the private attribute access would not be surfaced to the UI

yes, that would be just one consequence of this.  Essentially _any_ usage of `warnings.warn` would not go to the UI.  Someone would need to explicitly use `notifications.show_warning` instead

See for example #3704, and #3214, both of which were caused by a warning in vispy or Qt or something, usually environment/version specific.  the kind of thing that should perhaps fail a test (if you're surfacing warnings as errors as we do), but probably don't belong in the GUI warnings.

**Comment by sofroniewn** [comment]
makes sense to me

**Comment by goanpeca** [comment]
Do we need extra voting time :) ?

**Comment by tlambert03** [comment]
Nah, this could happen
